### PR TITLE
v4 exome release HT generation

### PR DESCRIPTION
**Pull request by mike-w-wilson**
There are still a few TODOs in here but figured I'd get this to you ASAP so you can start on it. We still need to update FAF once we have it and I also have to check with Julia on a couple of things to make sure we have the right files.

### CODE REVIEW

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py [hunk 1]
```diff
@@ -0,0 +1,668 @@
...
+            table,
+            _intervals=partition_intervals,
+            test=test,
+            release_exists=release_exists,
+        )
+        for table in tables
+    ]
+    joined_ht = reduce((lambda joined_ht, ht: joined_ht.join(ht, "left")), hts)
+
+    # Track the dataset we've added as well as the source path.
```
Comment (ch-kr) [inline gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:526]:
```suggestion
    # Track the datasets we've added as well as their source paths.
```

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py [hunk 2]
```diff
@@ -0,0 +1,668 @@
...
+    # If release in tables, read in the included datasets json
```
Comment (ch-kr) [inline gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:527]:
```suggestion
    # If release HT is included in tables, read in the included datasets json
```

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py [hunk 3]
```diff
@@ -0,0 +1,668 @@
...
+                hl.parse_locus_interval(
+                    "chr1:55039447-55064852", reference_genome="GRCh38"
+                )
+            ],
+        )
+    partition_intervals = base_ht._calculate_new_partitions(
+        base_ht.n_partitions() * new_partition_percent
+    )
+
+    # Reorg list so base table is first.
```
Comment (ch-kr) [inline gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:510]: this assumes that the base table is part of the list of tables to join, which it should be by default, maybe it's worth adding a warning / check or note to the docstring about this?

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py [hunk 4]
```diff
@@ -0,0 +1,668 @@
...
+    ht = join_hts(
+        args.base_table,
+        args.tables_for_join,
+        args.new_partition_percent,
+        args.test,
+        args.release_exists,
+    )
+
+    ht = hl.filter_intervals(ht, [hl.parse_locus_interval("chrM")], keep=False)
+    ht = ht.filter(hl.is_defined(ht.filters))
```
Comment (ch-kr) [inline gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:568]: naive question, but is this to make sure we're removing any AS_lowqual sites?

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py [hunk 5]
```diff
@@ -0,0 +1,668 @@
...
+
+    ht = ht.select_globals(**get_select_global_fields(ht))
+
+    ht = ht.annotate_globals(
+        filtering_model=ht.filtering_model.drop("model_id"),
+        vep_globals=ht.vep_globals.annotate(
+            gencode_version="Release 39",
+            mane_select_version="v0.95",
+        ),
+        tool_versions=ht.tool_versions.annotate(
```
Comment (ch-kr) [inline gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:578]: same question about keeping these values as constants somewhere (rather than hardcoding)

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py [hunk 6]
```diff
@@ -0,0 +1,668 @@
...
+        version=args.version,
+    )
+
+    # Reorder fields to match final schema.
+    ht = ht.select(*FINALIZED_SCHEMA["rows"]).select_globals(
+        *FINALIZED_SCHEMA["globals"]
+    )
+
+    output_path = (
+        qc_temp_prefix() + "release/gnomad.exomes.sites.test.ht"
```
Comment (ch-kr) [inline gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:600]: not important, but why not use an f-string here instead of the plus?

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py (hunk 4 above)
Comment (mike-w-wilson) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:627]: Yup!

**Comment by mike-w-wilson** [comment]
Back to you @ch-kr !

**Review by ch-kr — COMMENTED** [review]
just a few more nitpicks and a question about only_het for @jkgoodrich, otherwise current code minus joint faf LGTM

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py [hunk 7]
```diff
@@ -0,0 +1,727 @@
...
+        "vep_globals",
+        "date",
+        "version",
+    ],
+    "rows": [
+        "freq",
+        "grpmax",
+        "faf",
+        "gen_anc_faf_max",
+        # "joint_freq",
```
Comment (KoalaQin) [inline gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:99]: Where do you put 'InbreedingCoeff'? Not after here?

Comment (ch-kr) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:99]: it's in the info struct, isn't it?

Comment (KoalaQin) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:99]: I thought we are dropping that one, and use the one from freq_ht, calculated from call stats.

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py [hunk 8]
```diff
@@ -0,0 +1,727 @@
...
+    format="%(asctime)s (%(name)s %(lineno)s): %(message)s",
+    datefmt="%m/%d/%Y %I:%M:%S %p",
+)
+logger = logging.getLogger("create_release_ht")
+logger.setLevel(logging.INFO)
+
+# Remove InbreedingCoeff from allele-specific fields because it is processed separately
+# from the other fields.
+AS_FIELDS = deepcopy(AS_FIELDS)
+AS_FIELDS.remove("InbreedingCoeff")
```
Comment (KoalaQin) [inline gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:57]: Wouldn't this be: AS_FIELDS = AS_FIELDS.remove("InbreedingCoeff")

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py (hunk 7 above)
Comment (ch-kr) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:99]: yeah, in the code below, Mike adds this annotation from the freq HT to the info struct

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py (hunk 8 above)
Comment (ch-kr) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:57]: wouldn't that set AS_FIELDS to None?

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py [hunk 9]
```diff
@@ -0,0 +1,670 @@
...
+    score_name = hl.eval(filters_ht.filtering_model.score_name)
+    filters_ht = filters_ht.transmute(**filters_ht.truth_sets)
+    filters = filters_ht[ht.key]
+    filters_info_fields = [
+        "singleton",
+        "transmitted_singleton",
+        "omni",
+        "mills",
+        "monoallelic",
+        "only_het",
```
[the remark this answers is not in the document]
Comment (jkgoodrich) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:375]: Request for Konrad: https://app.zenhub.com/workspaces/gnomad-5f4d127ea61afc001d6be50b/issues/gh/broadinstitute/gnomad_production/705

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py (hunk 7 above)
Comment (KoalaQin) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:99]: ah Okay, got it!

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py (hunk 8 above)
Comment (KoalaQin) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:57]: you're right, sorry.

Comment (mike-w-wilson) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:57]: mutable types will get ya!

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py [hunk 10]
```diff
@@ -0,0 +1,727 @@
...
+    "filters",
+    "freq",
+    "info",
+    "region_flags",
+    "in_silico",
+    "vep",
+    # "joint_faf",
+]
+
+INSILICO_PREDICTORS = ["cadd", "spliceai", "pangolin", "revel", "phylop"]
```
Comment (KoalaQin) [inline gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:70]: a minor thing: I saw you have a different order below, could you make it consistent? 
`INSILICO_PREDICTORS = ["cadd", "revel", "spliceai", "pangolin", "phylop"]`

Comment (mike-w-wilson) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:70]: Yup that makes sense!

**Review by KoalaQin — CHANGES_REQUESTED** [review]

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py [hunk 11]
```diff
@@ -0,0 +1,668 @@
...
+            ),
+            "path": [
+                get_insilico_predictors(predictor=predictor).path
+                for predictor in INSILICO_PREDICTORS
+            ],
+            "field_name": "in_silico_predictors",
+            "select": [
+                "cadd",
+                "revel_max",
+                "spliceai_ds_max",
```
[the remark this answers is not in the document]
Comment (KoalaQin) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:180]: I changed it the schema. Thanks!

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py [hunk 12]
```diff
@@ -0,0 +1,728 @@
...
+        "freq_meta_sample_count",
+        "faf_meta",
+        "faf_index_dict",
+        # "joint_freq_meta",
+        # "joint_freq_index_dict",
+        # "joint_freq_meta_sample_count",
+        # "joint_faf_meta",
+        # "joint_faf_index_dict",
+        "age_distribution",
+        "downsamplings",
```
[the remark this answers is not in the document]
Comment (mike-w-wilson) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:85]: There is downsamplings for gnomad and for the non_ukb subset

[the remark this answers is not in the document]
Comment (ch-kr) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:85]: oh oops, thanks!

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py [hunk 13]
```diff
@@ -0,0 +1,728 @@
...
+        "filtering_model",
+        "inbreeding_coeff_cutoff",
+        "tool_versions",
+        "vrs_versions",
+        "vep_globals",
+        "date",
+        "version",
+    ],
```
Comment (ch-kr) [inline gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:93]: it just occurred to me, if we are adding interval QC as a region flag, should we add the interval QC params to the globals?

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py [hunk 14]
```diff
@@ -0,0 +1,728 @@
...
+    "rows": [
+        "freq",
+        "grpmax",
+        "faf",
+        "gen_anc_faf_max",
```
Comment (ch-kr) [inline gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:98]: I was confused about the difference between the `faf95` field within the `grpmax` struct and the `faf95_max` field in this `gen_anc_faf_max` struct. From chatting with Kristen, I learned that `grpmax.faf95` is the FAF of the gen anc group with the highest **AF** for a variant, and the `gen_anc_faf_max.faf95_max` is the FAF of the gen anc group with the highest **FAF** for a variant. These values are not necessarily the same; the group with the highest AF for a variant isn't necessarily the group with the highest FAF for the variant.

I then asked Anne if there is any utility for the `grpmax.faf95`, and she said no -- I vote we remove `grpmax.faf95` from our release HTs

Comment (KoalaQin) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:98]: I agree, it was very confusing when I was writing the `gen_anc_faf_max` function.

Comment (mike-w-wilson) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:98]: Should we rename it to fafmax? I could just do this in this release code for now

Comment (ch-kr) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:98]: I like this, particularly since this struct contains fields noting which gen anc group the FAF95/FAF99 values are from

Comment (mike-w-wilson) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:98]: Just to be sure, this == `gen_anc_faf_max` ?

Comment (ch-kr) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:98]: yes: I like your proposal of renaming `gen_anc_faf_max` to simply `fafmax`, particularly since `gen_anc_faf_max` already contains fields like `faf95_max_gen_anc` which indicate the genetic ancestry group associated with the fafmax values.

tl;dr -- unless anyone has strong opinions, let's rename `gen_anc_faf_max` to `fafmax` and drop `faf95` from `grpmax`

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py (hunk 13 above)
Comment (mike-w-wilson) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:93]: These are the globals on interval_pass_qc().ht()
```
Global fields:
    'high_qual_interval_parameters': struct {
        per_platform: bool, 
        all_platforms: bool, 
        high_qual_cutoffs: dict<str, array<tuple (
            str, 
            str, 
            float64
        )>>, 
        min_platform_size: int32
    } 
```
Is this what you had in mind?

Comment (ch-kr) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:97]: yes, thank you!

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py [hunk 15]
```diff
@@ -245,7 +246,8 @@ def get_config(
         "joint_faf": {
             "ht": get_combined_faf_release().ht(),
             "path": get_combined_faf_release().path,
-            "select": ["joint_freq", "joint_grpmax", "joint_faf", "joint_fafmax"],
+            "custom_select": custom_joint_faf_select,
+            "select": ["joint_freq", "joint_faf", "joint_fafmax"],
```
Comment (KoalaQin) [inline gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:250]: Why do you make "custom_select" before "select" here?

Comment (KoalaQin) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:250]: Another question: will the order change? Does it matter?

Comment (mike-w-wilson) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:249]: I flipped it back but the order of the keys in the config shouldnt matter but I do agree that consistency is nice here. The annotation order will not change because it goes by the order in the `FINALIZED_SCHEMA` variable.

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py [hunk 16]
```diff
@@ -0,0 +1,761 @@
...
+    # Create a dict of the fields from the freq HT that we want to add to the info.
+    freq_ht = get_config().get("freq")["ht"]
+    freq_info_dict = {"inbreeding_coeff": freq_ht[ht.key]["inbreeding_coeff"]}
+
+    # Create a dict of the fields from the VRS HT that we want to add to the info.
+    vrs_ht = get_vrs().ht()
+    vrs_info_fields = {"vrs": vrs_ht[ht.key].vrs}
+
+    # Create a dict of the fields from the info HT that we want keep in the info.
+    info_struct = hl.struct(**ht.site_info, **ht[f"{compute_info_method}_info"])
```
Comment (KoalaQin) [inline gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:418]: Here `ht` still means `get_info().ht()`, right? Does it have `site_info`?

Comment (mike-w-wilson) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:418]: ht here is `hl.read_table(ht_path, _intervals=_intervals)` from the get_ht function. ht_path would be `get_info().path`. It's basically equivalent to get_info().ht() but its read in with the same intervals as freq to improve the join efficiency.  Yes it does have a `site_info`.

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py [hunk 17]
```diff
@@ -268,6 +269,37 @@ def get_config(
...
+    return selects
+
+
+def custom_freq_select(ht: hl.Table) -> Dict[str, hl.expr.Expression]:
+    """
+    Drop faf95 from both 'gnomad' and 'non_ukb' in 'grpmax' and rename `gen_anc_faf_max` to `fafmax`.
+
+    These annotations will be combined with the others from freq's select in the config.
+
+    :param ht: Freq Hail Table.
```
Comment (ch-kr) [inline gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:292]:
```suggestion
    These annotations will be combined with the others from freq's select in the config.

    .. note::
        - The faf95 field in the grpmax struct is the FAF of the genetic ancestry group with the largest AF (grpmax AF).
        - The FAF fields within the gen_anc_faf_max struct contains the FAFs from the genetic ancestry group(s) with the largest FAFs
        - These values aren't necessarily the same; the group with the highest AF for a variant isn't necessarily the group with the highest FAF for a variant
        - The filtering allele frequencies that are used by the community are the values within the gen_anc_faf_max struct, NOT grpmax FAF, which is why we are dropping grpmax.faf95 and renaming gen_anc_faf_max
        
    :param ht: Freq Hail Table
```

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py [hunk 18]
```diff
@@ -268,6 +269,37 @@ def get_config(
     return config
 
 
+def custom_joint_faf_select(ht: hl.Table) -> Dict[str, hl.expr.Expression]:
+    """
+    Drop faf95 from 'grpmax'.
+
+    This annotations will be combined with the others from joint_faf's select in the config.
```
Comment (ch-kr) [inline gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:276]:
```suggestion
    This annotation will be combined with the others from joint_faf's select in the config.
    See note in `custom_freq_select` explaining why this field is removed.
```

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py [hunk 19]
```diff
@@ -64,7 +68,7 @@
     "region_flags",
     "in_silico",
     "vep",
-    # "joint_faf",
+    "joint_faf",
```
Comment (ch-kr) [inline gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:71]: another global I realized was missing from this HT while documenting the schema on the help/faq page is a potential readme describing how to use the frequency fields (this was proposed for v3) -- is this something we still want to add, or will we point users to the instructions on the help page?

**Review by ch-kr — COMMENTED** [review]
two minor documentation suggestions and one question

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py (hunk 19 above)
Comment (mike-w-wilson) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:71]: Ah right, Dan King asked for that right? I think it makes sense to include if that got all the way back to us from hail users. It's not super intuitive. I'll draft something.

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py [hunk 20]
```diff
@@ -658,9 +658,9 @@ def main(args):
         args.release_exists,
     )
 
-    # Filter out chrM and AS_lowqual sites.
+    # Filter out chrM, AS_lowqual sites, and AC_raw == 0.
     ht = hl.filter_intervals(ht, [hl.parse_locus_interval("chrM")], keep=False)
-    ht = ht.filter(hl.is_defined(ht.filters))
+    ht = ht.filter(hl.is_defined(ht.filters) & ht.freq[1].AC > 0)
```
Comment (KoalaQin) [inline gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py]: Shouldn't it be `ht = ht.filter(hl.len(ht.filters)==0) & (ht.freq[1].AC > 0))`? Also `AS_lowqual` seems to be from get_info() and we didn't select, no? `filters` is from `final_filter().ht()`, right?

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py (hunk 19 above)
Comment (jkgoodrich) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:71]: wait, isn't that here: https://gnomad.broadinstitute.org/help#frequently-asked-questions under How do I access the gnomAD Hail Table frequency annotation

Comment (jkgoodrich) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:71]: It think we just point there right?

Comment (jkgoodrich) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:71]: Maybe put that as a main topic not just a frequently asked question?

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py (hunk 20 above)
Comment (KoalaQin) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py]: oh okay, but you need () for (ht.freq[1].AC > 0), otherwise there's an TypeError.

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py (hunk 19 above)
Comment (ch-kr) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:71]: yes, it's on the help page, but it doesn't hurt to add to the HT itself as well, right?

here's Dan's initial request
>Random thought, maybe you can include a README global field in the next release with just some text describing where to get more information about the meaning of adj, raw, what the freq fields are, etc. Or maybe there could be a more visible link to the Help section right about the sites table? I'm not exactly sure what the issue is but maybe a couple times a year someone shows up to the Hail forum flummoxed by the freq array

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py (hunk 20 above)
Comment (mike-w-wilson) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py]: Yes thanks thats been updated

File: gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py (hunk 19 above)
Comment (mike-w-wilson) [reply gnomad_qc/v4/create_release/create_release_sites_ht_exomes.py:72]: Added and pointed to the FAQ https://github.com/broadinstitute/gnomad_qc/pull/477/commits/b3b9fcefc0841a2cce618b39b5cb98138125c70d

File: gnomad_qc/v4/resources/release.py [hunk 21]
```diff
@@ -19,6 +19,46 @@
     RELEASES,
 )
 
+FREQUENCY_README = """
```
Comment (mike-w-wilson) [inline gnomad_qc/v4/resources/release.py:22]: @ch-kr Added this but welcome your wordsmithing

Comment (mike-w-wilson) [reply gnomad_qc/v4/resources/release.py:22]: One note, when this is hl.eval'd, it looks .....bad because eval just translates special characters to have escape sequences...I couldnt find a way around this, it will be on the user to parse

File: gnomad_qc/v4/resources/release.py [hunk 22]
```diff
@@ -19,6 +19,46 @@
...
+The 'freq' row annotation contains allele frequency information aggregated by certain
+sample groupings within the dataset within an array. Each element of the array contains
+a grouping's alternate allele count (AC), alternate allele frequency (AF), total number
+ofalleles (AN), and the count of homozygous alternate individuals (homozygote_count).
```
Comment (ch-kr) [inline gnomad_qc/v4/resources/release.py:26]:
```suggestion
The 'freq' row annotation is an array that contains allele frequency information. Each element of the array is a struct that contains the alternate allele count (AC), alternate allele frequency (AF), total number of alleles (AN), and number of homozygous alternate individuals (homozygote_count) for a specific sample grouping.
```
now I'm wondering if I should reword the help/faq as well

File: gnomad_qc/v4/resources/release.py (hunk 21 above)
Comment (ch-kr) [reply gnomad_qc/v4/resources/release.py:22]: running
```
print(hl.eval(ht.readme), sep=os.linesep)
```
helped! it's still a bit ugly but no longer has `\n`s everywhere

File: gnomad_qc/v4/resources/release.py [hunk 23]
```diff
@@ -19,6 +19,46 @@
...
+
+To retrieve the frequency information for a specific group of samples, use the
+'freq_index_dict' global annotation. This annotation is a dictionary keyed by sample
+grouping combinations whose values are the combination's index in the 'freq' array.
```
Comment (ch-kr) [inline gnomad_qc/v4/resources/release.py:30]:
```suggestion
Use the 'freq_index_dict' global annotation to retrieve frequency information for a specific group of samples from the 'freq' array. This global annotation is a dictionary keyed by sample
grouping combinations whose values are the combination's index in the 'freq' array.
```

**Review by ch-kr — APPROVED** [review]
minor rewording suggestion but LGTM!

**Review by KoalaQin — APPROVED** [review]
LGTMT! Bravo!
